File: android/messenger-sdk/src/test/kotlin/com/dixa/messenger/FakeTransport.kt
package com.dixa.messenger

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

/**
 * In-memory [Transport] for tests. No real socket, no timers — you drive it by
 * hand so tests are deterministic under the coroutine test scheduler.
 *
 * Typical use:
 * ```
 * val transport = FakeTransport()
 * val scope = CoroutineScope(UnconfinedTestDispatcher(testScheduler))
 * val messenger = Messenger(MessengerConfig.Builder().build(), transport, scope)
 * messenger.setListener(listener)
 * messenger.connect()                 // FakeTransport emits Connecting then Connected;
 *                                      // the unconfined dispatcher runs collectors eagerly
 * transport.pushInbound("""{"type":"message", ... }""")
 * ```
 */
class FakeTransport : Transport {

    private val inbound = MutableSharedFlow<String>(replay = 0, extraBufferCapacity = 64)
    private val state = MutableSharedFlow<ConnectionState>(replay = 1, extraBufferCapacity = 8)

    override val inboundMessages: SharedFlow<String> = inbound.asSharedFlow()
    override val connectionState: SharedFlow<ConnectionState> = state.asSharedFlow()

    /** Frames handed to [send], in order — assert your outbound wire format here. */
    val sentFrames: MutableList<String> = mutableListOf()

    /** How many times [connect] was called. */
    var connectCallCount: Int = 0
        private set

    override fun connect() {
        connectCallCount++
        state.tryEmit(ConnectionState.Connecting)
        state.tryEmit(ConnectionState.Connected)
    }

    override fun disconnect() {
        state.tryEmit(ConnectionState.Disconnected)
    }

    override fun send(text: String) {
        sentFrames.add(text)
    }

    // ---- Test controls ----

    /** Deliver a raw inbound JSON frame to every live collector. */
    fun pushInbound(rawJson: String) {
        inbound.tryEmit(rawJson)
    }

    /** Simulate the socket dropping (bad signal, app backgrounded, etc.). */
    fun simulateDisconnect() {
        state.tryEmit(ConnectionState.Disconnected)
    }

    /** Simulate the socket coming back after a drop. */
    fun simulateReconnect() {
        state.tryEmit(ConnectionState.Connecting)
        state.tryEmit(ConnectionState.Connected)
    }
}

File: ios/Tests/DixaMessengerTests/FakeTransport.swift
import Foundation
import Combine
@testable import DixaMessenger

/// In-memory `Transport` for tests. No real socket, no timers — you drive it by
/// hand so tests are deterministic.
///
/// Typical use:
/// ```
/// let transport = FakeTransport()
/// let messenger = DixaMessenger(configuration: DixaConfiguration(), transport: transport)
/// messenger.onMessage = { ... }
/// messenger.connect()                 // FakeTransport emits .connecting then .connected
/// transport.pushInbound(#"{"type":"message", ... }"#)
/// ```
final class FakeTransport: Transport {

    private let inboundSubject = PassthroughSubject<String, Never>()
    private let stateSubject = PassthroughSubject<ConnectionState, Never>()

    /// Frames handed to `send(_:)`, in order — assert your outbound wire format here.
    private(set) var sentFrames: [String] = []
    /// How many times `connect()` was called.
    private(set) var connectCallCount = 0

    var inboundMessages: AnyPublisher<String, Never> {
        inboundSubject.eraseToAnyPublisher()
    }

    var connectionState: AnyPublisher<ConnectionState, Never> {
        stateSubject.eraseToAnyPublisher()
    }

    func connect() {
        connectCallCount += 1
        stateSubject.send(.connecting)
        stateSubject.send(.connected)
    }

    func disconnect() {
        stateSubject.send(.disconnected)
    }

    func send(_ text: String) {
        sentFrames.append(text)
    }

    // MARK: - Test controls

    /// Deliver a raw inbound JSON frame to the SDK.
    func pushInbound(_ rawJSON: String) {
        inboundSubject.send(rawJSON)
    }

    /// Simulate the socket dropping (bad signal, backgrounded, etc.).
    func simulateDisconnect() {
        stateSubject.send(.disconnected)
    }

    /// Simulate the socket coming back after a drop.
    func simulateReconnect() {
        stateSubject.send(.connecting)
        stateSubject.send(.connected)
    }
}
